design/design-doc-custom-composition.md

```diff
@@ -0,0 +1,179 @@
+# Custom Composition Types
+
+* Owner: Muvaffak Onuş (@muvaf)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Background
+
+## Proposal
+
+### Authorship Experience
+
+Let's say we have the following `CompositeResourceDefinition` that defines a `PrivateNetwork`
+type.
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: InfrastructureDefinition
+metadata:
+  name: privatenetworks.aws.example.org
+spec:
+  crdSpecTemplate:
+    group: aws.example.org
+    version: v1alpha1
+    names:
+      kind: PrivateNetwork
+      listKind: PrivateNetworkList
+      plural: privatenetworks
+      singular: privatenetwork
+    validation:
+      openAPIV3Schema:
+        type: object
+        properties:
+          spec:
+            type: object
+            properties:
+              region:
+                type: string
+                description: Geographic region in which nodes will exist.
+              providerConfigRef:
+                type: object
+                description: Crossplane AWS provider credentials to use.
+                properties:
+                  name:
+                    type: string
+                required:
+                - name
+            required:
+            - region
+            - providerConfigRef
+```
+
+User can have multiple `Composition` or custom compositions that target this new
+type. Let's assume they would like to write a cdk8s app and use it as the composition
+engine. They'd need to write the cdk8s app that will take the following resources
+as input:
+  * `PrivateNetwork` custom resource.
+  * `CustomComposition` custom resource.
+  It will return all the resulting resources as output.
+
+This conversion function will be wrapped by an HTTP server and pushed as an image
+whose `ENTRYPOINT` is to start that HTTP server. The input will come as POST
+request and the body of the result will consist of all the output resources.
+
+At this point, we have our cdk8s app that will act as composition engine. Now,
+let's go through how the consumers of that app will be able to use it.
+
+### Consumption Experience
+
+Now that our composition engine is ready, we'll instruct Crossplane to install
+that composition engine and send the appropriate requests. In order to do that,
+the following `CustomComposition` will be created:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: CustomComposition
+metadata:
+  name: acmenetwork.acme.org
+spec:
+  writeConnectionSecretsToNamespace: crossplane-system
+  reclaimPolicy: Delete
+  compositeTypeRef:
+    apiVersion: aws.example.org/v1alpha1
+    kind: PrivateNetwork
+  image: acme/network-cdk8s:1.0.0
```

Comment:
I wonder if (eventually) we would do anything to validate that this image we're running is in fact serving custom composition logic. We probably don't want to become a vector to deploy arbitrary containers.

Reply:
Honestly, that's the part I'm struggling with most. I am hesitant to add yet another reason for Crossplane to be super user. One option is to have the user create the `Deployment` alongside `CustomComposition`:
```yaml
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: CustomComposition
metadata:
  name: acmenetwork.acme.org
spec:
  writeConnectionSecretsToNamespace: crossplane-system
  reclaimPolicy: Delete
  compositeTypeRef:
    apiVersion: aws.example.org/v1alpha1
    kind: PrivateNetwork
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: acmenetwork.acme.org
  labels:
    app: acmenetwork.acme.org
spec:
  replicas: 1
  selector:
    matchLabels:
      app: acmenetwork.acme.org
  template:
    metadata:
      labels:
        app: acmenetwork.acme.org
    spec:
      containers:
      - name: custom-composition
        image: crossplane/custom-composition:0.0.1
        env:
        # COMPOSITE_TYPE will let the custom-composition controller know what type
        # it should watch.
        - name: COMPOSITE_TYPE
          value: privatenetworks.aws.example.org
        # CUSTOMCOMPOSITION_NAME is the name of the CustomComposition that will
        # be sent over to composition engine alongside composite instance. 
        - name: CUSTOMCOMPOSITION_NAME
          value: acmenetwork.acme.org
      - name: composition-engine
        image: acme/network-cdk8s:1.0.0
        ports:
        - containerPort: 80
```

`CustomComposition` wouldn't have any controllers and its only goal would be to be referenced so that composite can demand this `CustomComposition`. It saves Crossplane from `Deployment` permission but UX is getting a hit since user has to write that `Deployment`.

> I wonder if (eventually) we would do anything to validate that this image we're running is in fact serving custom composition logic. We probably don't want to become a vector to deploy arbitrary containers.

I'm not sure if we could. In fact, the `Deployment` needs a service account with capabilities to create the output of the cdk8s app and there isn't much way for Crossplane to know what cdk8s needs until actual creation.

Thinking through this while writing, I think that might be the right approach. The full YAML that cdk8s can print could have:
* 1 `CompositeResourceDefinition` (optional)
* 1 `CustomComposition`
* 1 `Deployment`
* 1 `ServiceAccount`, 1 `ClusterRole`, 1 `ClusterRoleBinding` (these will need to be created by the user in all cases, otherwise Crossplane would have to give `cluster-admin` role to the `Deployment`)

(having a spider-sense saying we'll land on making this a package, probably a `Configuration` or a new type)

@negz what's your take?

Reply:
I have given this some more thought. To some extent, this is very similar to a provider in the sense that it sees a type and creates some objects in an API, though API in this case is Kubernetes and the created objects are not limited to CRDs it brings with itself.

So, an option alternative to the earlier [suggestion](https://github.com/crossplane/crossplane/pull/1705/files#r483710125) is that we can have a type, like `ExternalComposition.pkg.crossplane.io` (not sure about the name), that will create the RBAC resources according to what's given in `crossplane.yaml` ([permission](https://github.com/crossplane/crossplane/issues/1767) requests to certain kinds it can manage) and also create the necessary `Deployment` with two containers (one is `external-composition-controller` the other one is the app itself). Users would choose the `ExternalComposition` in their `compositionRef`, Crossplane itself wouldn't reconcile it at all and the `external-composition-controller` would reconcile and call the endpoints of the app running as sidecar container.

The overall DX using cdk8s would be that I write some JavaScript, a `crossplane.yaml` for metadata and permissions and a `Dockerfile`. Then `build` it similar to provider builds & push (image path is given in `crossplane.yaml`). Crossplane package manager will know what operations to do when it deals with this `kind` of package.

In the earlier [suggestion](https://github.com/crossplane/crossplane/pull/1705/files#r483710125) we can have the CLI generate those extra YAMLs and let user do the `kubectl apply`. Developer could make sure that each build runs that generation and package them up in a Helm chart for external usage, similar to controllers out in the wild. But instead, if we do the generation in Package Manager and use the standard Crossplane packaging format, it's less cumbersome while we take on all the responsibility regarding RBAC management of that package. The main difference between providers and this package would be that the final `Deployment` is controller+server and that may require some customization.

I feel closer to a new packaging type, it feels like even if we do start with generation of those YAMLs using CLI we'll end up deciding to make a new package type since the overall operation is pretty similar to what package manager does. But I'm not yet fully sold on the idea.

What do you guys think? Is this feasible at all and is it a good API to interact with? @hasheddan @negz @kasey @prasek

Reply:
@muvaf I am thinking along the same lines as you and I see a lot of parallels to some of the work in https://github.com/crossplane-contrib/crisscross. The difference between `crisscross` and this scenario is that we can't use generic controller logic (i.e. the managed reconciler) and just define certain endpoints when the controller reaches out for the custom logic. However, I would like to explore if we could bring these two ideas closer together and potentially implement a single package type that satisfies both (right now I am thinking something along the lines of `Function.pkg.crossplane.io`.

We could go about bringing the two ideas together by encoding more of the reconcile logic into the functions themselves, and, in the case of crisscross, potentially abstracting that away with a library / base image so users still _feel_ like they are just writing endpoints. In Crossplane, this would look like dynamically spinning up controllers that only make an http request to a the function, where all logic is encompassed.

As you mentioned, installing a `Function` would allow us to make sure the appropriate RBAC is provided, as well as setup certificates for mutual TLS, etc. What do you think about this?